quickjs.c

```diff
@@ -1440,7 +1443,17 @@ static void js_trigger_gc(JSRuntime *rt, size_t size)
                    (uint64_t)rt->malloc_state.malloc_size);
         }
 #endif
-        JS_RunGC(rt);
+        //To ensure JS_RunGC cannot be executed again within callbacks, disable and restore it after.
+        size_t tmp_threshold = rt->malloc_gc_threshold;
+        rt->malloc_gc_threshold=-1;
+
+        if((rt->hooks == NULL) || rt->hooks(JS_HOOK_GC_BEFORE,NULL)){
```

Comment:
The hook could change the threshold, but you are then re-setting it.

Reply:
Yes, but while functionally similar they have different applications. Setting the GC on/off in specific critical regions can get a bit tricky with async/await. Not only that, but doing so would mask that a request was done, preventing the application to say "oh yes, I had to delay it, I am now in a safe region, let's clean up things". Ensuring the `before` event can say, "hey no, please don't" is a fail-safe. Ideally the custom callback could set up alternative mechanism to run that gc event, just a bit later when no critical task it taking place.

Reply:
Sounds very theoretical ;-) Why can't the mechanism that _knows_ if we are safe or not enable / disable GC while in a critical region?

Reply:
Based on the application requirements:
- Option 1: I don't care, gc events as needed and scheduled by the engine. Not good for real time applications.
- Option 2: gc events only when I determine regions are safe (possibly never). The only feasible approach for the strictest real time applications.
- Option 3: I would prefer gc to take place, but I cannot do it right now. I take notice of that, and skip it for now. In the meanwhile I ask the system to move itself in a safe state when possible, schedule the gc event in that safe state, and recover the current job from where it was left.

Having the option to skip gc events while tracking them enables option 3 which is better than option 2 for a good class of real time applications.  
It is not theoretical, it is literally how my application works.  
It does not mean it is good design, just that at the very least it is slightly above a purely theoretical scenario.

Reply:
Thanks for elaborating! Looks like option 3 could work as follows:

- GC event hits
- app checks if it can afford it, if not it disabled GC and tries to move to the safe state
- Once the safe state is hit, enable GC and run it
- goto 1

Wouldn't that work?

Reply:
By the time the app can check if it is ok to run a GC event, it is already in the body of `before`.  
If `before` does not stop it by failing what does?  
Unless you plan to split `before` and `evaluate` in two separate hooks, but it would be virtually identical just for more complexity.

Reply:
Before can stop GC from running by disabling it.

So before get hit, but because you disable it it doesn't run. Then when you want GC the app can enable it and run it.

What wouldn't work?